File: src/git/formatter.ts
import chalk from 'chalk';
import { ParsedDiff } from './parser';
import type { ParsedLine } from './parser';

export type OutputFormat = 'terminal' | 'markdown' | 'json';

export interface FormatterOptions {
    showLineNumbers?: boolean;
    showStats?: boolean;
    contextLines?: number;
}

export class DiffFormatter {
    /**
     * Format diff for terminal display with colors
     */
    toTerminal(diffs: ParsedDiff[], options: FormatterOptions = {}): string {
        const { showLineNumbers = true, showStats = true } = options;
        const output: string[] = [];

        for (const diff of diffs) {
            // File header
            output.push(chalk.bold.white('─'.repeat(60)));
            output.push(this.formatFileHeader(diff));

            if (showStats) {
                output.push(this.formatStats(diff.additions, diff.deletions));
            }
            output.push('');

            // Hunks
            for (const hunk of diff.hunks) {
                output.push(chalk.cyan(hunk.header));

                for (const line of hunk.lines) {
                    output.push(this.formatLine(line, showLineNumbers));
                }
                output.push('');
            }
        }

        return output.join('\n');
    }

    /**
     * Format file header with appropriate color
     */
    private formatFileHeader(diff: ParsedDiff): string {
        if (diff.isNew) {
            return chalk.green.bold(`+ New: ${diff.newFile}`);
        } else if (diff.isDeleted) {
            return chalk.red.bold(`- Deleted: ${diff.oldFile}`);
        } else if (diff.isRenamed) {
            return chalk.yellow.bold(`→ Renamed: ${diff.oldFile} → ${diff.newFile}`);
        } else {
            return chalk.blue.bold(`Modified: ${diff.newFile}`);
        }
    }

    /**
     * Format stats line
     */
    private formatStats(additions: number, deletions: number): string {
        const addStr = chalk.green(`+${additions}`);
        const delStr = chalk.red(`-${deletions}`);
        return `  ${addStr} ${delStr}`;
    }

    /**
     * Format a single diff line with color
     */
    private formatLine(line: ParsedLine, showLineNumbers: boolean): string {
        let lineNum = '';

        if (showLineNumbers) {
            const oldNum = line.oldLineNumber?.toString().padStart(4, ' ') || '    ';
            const newNum = line.newLineNumber?.toString().padStart(4, ' ') || '    ';
            lineNum = chalk.gray(`${oldNum} ${newNum} │ `);
        }

        const prefix = line.type === 'add' ? '+' : line.type === 'delete' ? '-' : ' ';
        const content = prefix + line.content;

        switch (line.type) {
            case 'add':
                return lineNum + chalk.green(content);
            case 'delete':
                return lineNum + chalk.red(content);
            case 'context':
                return lineNum + chalk.gray(content);
        }
    }

    /**
     * Format diff as markdown (for LLM consumption)
     */
    toMarkdown(diffs: ParsedDiff[], options: FormatterOptions = {}): string {
        const { showStats = true } = options;
        const output: string[] = [];

        output.push('# Git Diff Summary\n');

        // Overall stats
        const totalAdditions = diffs.reduce((sum, d) => sum + d.additions, 0);
        const totalDeletions = diffs.reduce((sum, d) => sum + d.deletions, 0);
        output.push(`**Files changed:** ${diffs.length}`);
        output.push(`**Additions:** +${totalAdditions} | **Deletions:** -${totalDeletions}\n`);

        for (const diff of diffs) {
            // File header
            const status = diff.isNew ? '(new)' : diff.isDeleted ? '(deleted)' : diff.isRenamed ? '(renamed)' : '';
            output.push(`## ${diff.newFile} ${status}`);

            if (showStats && (diff.additions > 0 || diff.deletions > 0)) {
                output.push(`*+${diff.additions} -${diff.deletions}*\n`);
            }

            // Code block with diff
            output.push('```diff');
            for (const hunk of diff.hunks) {
                output.push(hunk.header);
                for (const line of hunk.lines) {
                    const prefix = line.type === 'add' ? '+' : line.type === 'delete' ? '-' : ' ';
                    output.push(prefix + line.content);
                }
            }
            output.push('```\n');
        }

        return output.join('\n');
    }

    /**
     * Format diff as JSON (for API responses)
     */
    toJSON(diffs: ParsedDiff[]): string {
        return JSON.stringify({
            summary: {
                files: diffs.length,
                additions: diffs.reduce((sum, d) => sum + d.additions, 0),
                deletions: diffs.reduce((sum, d) => sum + d.deletions, 0),
            },
            files: diffs,
        }, null, 2);
    }

    /**
     * Create a compact summary string
     */
    toSummary(diffs: ParsedDiff[]): string {
        const files = diffs.length;
        const additions = diffs.reduce((sum, d) => sum + d.additions, 0);
        const deletions = diffs.reduce((sum, d) => sum + d.deletions, 0);

        const fileList = diffs.map(d => {
            const status = d.isNew ? '+ ' : d.isDeleted ? '- ' : d.isRenamed ? '→ ' : 'M ';
            return `${status}${d.newFile}`;
        });

        return `${files} file(s) changed, +${additions} -${deletions}\n\n${fileList.join('\n')}`;
    }
}

File: go-source/internal/git/formatter.go
package git

import (
	"encoding/json"
	"fmt"
	"strings"

	"github.com/fatih/color"
)

type FormatterOptions struct {
	ShowLineNumbers bool
	ShowStats       bool
}

type DiffFormatter struct{}

func NewDiffFormatter() *DiffFormatter { return &DiffFormatter{} }

func (f *DiffFormatter) ToTerminal(diffs []ParsedDiff, options FormatterOptions) string {
	showLineNumbers := true
	showStats := true
	if options.ShowLineNumbers == false {
		showLineNumbers = false
	}
	if options.ShowStats == false {
		showStats = false
	}

	out := make([]string, 0)
	for _, diff := range diffs {
		out = append(out, color.New(color.Bold).Sprint(strings.Repeat("─", 60)))
		out = append(out, f.formatFileHeader(diff))
		if showStats {
			out = append(out, fmt.Sprintf("  %s %s", color.GreenString("+%d", diff.Additions), color.RedString("-%d", diff.Deletions)))
		}
		out = append(out, "")
		for _, h := range diff.Hunks {
			out = append(out, color.CyanString(h.Header))
			for _, line := range h.Lines {
				out = append(out, f.formatLine(line, showLineNumbers))
			}
			out = append(out, "")
		}
	}
	return strings.Join(out, "\n")
}

func (f *DiffFormatter) formatFileHeader(diff ParsedDiff) string {
	switch {
	case diff.IsNew:
		return color.New(color.FgGreen, color.Bold).Sprintf("+ New: %s", diff.NewFile)
	case diff.IsDeleted:
		return color.New(color.FgRed, color.Bold).Sprintf("- Deleted: %s", diff.OldFile)
	case diff.IsRenamed:
		return color.New(color.FgYellow, color.Bold).Sprintf("→ Renamed: %s → %s", diff.OldFile, diff.NewFile)
	default:
		return color.New(color.FgBlue, color.Bold).Sprintf("Modified: %s", diff.NewFile)
	}
}

func (f *DiffFormatter) formatLine(line ParsedLine, showLineNumbers bool) string {
	lineNum := ""
	if showLineNumbers {
		oldNum := "    "
		newNum := "    "
		if line.OldLineNumber != nil {
			oldNum = fmt.Sprintf("%4d", *line.OldLineNumber)
		}
		if line.NewLineNumber != nil {
			newNum = fmt.Sprintf("%4d", *line.NewLineNumber)
		}
		lineNum = color.HiBlackString("%s %s │ ", oldNum, newNum)
	}
	prefix := " "
	if line.Type == LineAdd {
		prefix = "+"
	}
	if line.Type == LineDelete {
		prefix = "-"
	}
	content := prefix + line.Content
	switch line.Type {
	case LineAdd:
		return lineNum + color.GreenString(content)
	case LineDelete:
		return lineNum + color.RedString(content)
	default:
		return lineNum + color.HiBlackString(content)
	}
}

func (f *DiffFormatter) ToMarkdown(diffs []ParsedDiff) string {
	out := make([]string, 0)
	out = append(out, "# Git Diff Summary", "")
	adds, dels := 0, 0
	for _, d := range diffs {
		adds += d.Additions
		dels += d.Deletions
	}
	out = append(out, fmt.Sprintf("**Files changed:** %d", len(diffs)))
	out = append(out, fmt.Sprintf("**Additions:** +%d | **Deletions:** -%d", adds, dels), "")

	for _, d := range diffs {
		status := ""
		if d.IsNew {
			status = "(new)"
		} else if d.IsDeleted {
			status = "(deleted)"
		} else if d.IsRenamed {
			status = "(renamed)"
		}
		out = append(out, fmt.Sprintf("## %s %s", d.NewFile, status))
		if d.Additions > 0 || d.Deletions > 0 {
			out = append(out, fmt.Sprintf("*+%d -%d*", d.Additions, d.Deletions), "")
		}
		out = append(out, "```diff")
		for _, h := range d.Hunks {
			out = append(out, h.Header)
			for _, line := range h.Lines {
				prefix := " "
				if line.Type == LineAdd {
					prefix = "+"
				}
				if line.Type == LineDelete {
					prefix = "-"
				}
				out = append(out, prefix+line.Content)
			}
		}
		out = append(out, "```", "")
	}
	return strings.Join(out, "\n")
}

func (f *DiffFormatter) ToJSON(diffs []ParsedDiff) string {
	payload := map[string]any{
		"summary": map[string]any{
			"files":     len(diffs),
			"additions": sumAdds(diffs),
			"deletions": sumDels(diffs),
		},
		"files": diffs,
	}
	b, _ := json.MarshalIndent(payload, "", "  ")
	return string(b)
}

func (f *DiffFormatter) ToSummary(diffs []ParsedDiff) string {
	files := len(diffs)
	adds, dels := sumAdds(diffs), sumDels(diffs)
	list := make([]string, 0, len(diffs))
	for _, d := range diffs {
		status := "M "
		if d.IsNew {
			status = "+ "
		} else if d.IsDeleted {
			status = "- "
		} else if d.IsRenamed {
			status = "→ "
		}
		list = append(list, status+d.NewFile)
	}
	return fmt.Sprintf("%d file(s) changed, +%d -%d\n\n%s", files, adds, dels, strings.Join(list, "\n"))
}

func sumAdds(diffs []ParsedDiff) int {
	t := 0
	for _, d := range diffs {
		t += d.Additions
	}
	return t
}

func sumDels(diffs []ParsedDiff) int {
	t := 0
	for _, d := range diffs {
		t += d.Deletions
	}
	return t
}
